test/Test/Database/LSMTree/Normal/StateMachine.hs

```diff
@@ -0,0 +1,696 @@
+          ks1 == ks2 && var1 == var2
+      go (RangeLookup range1 var1)  (RangeLookup range2 var2) =
+          range1 == range2 && var1 == var2
+      go (Updates ups1 var1)        (Updates ups2 var2) =
+          Just ups1 == cast ups2 && Just var1 == cast var2
+      go (Inserts inss1 var1)       (Inserts inss2 var2) =
+          Just inss1 == cast inss2 && Just var1 == cast var2
+      go (Deletes ks1 var1)         (Deletes ks2 var2) =
+          Just ks1 == cast ks2 && Just var1 == cast var2
+      go (RetrieveBlobs var1 vars1) (RetrieveBlobs var2 vars2) =
+          Just var1 == cast var2 && Just vars1 == cast vars2
+      go (Snapshot name1 var1)      (Snapshot name2 var2) =
+          name1 == name2 && Just var1 == cast var2
+      go (Open name1)               (Open name2) =
+          name1 == name2
+      go (Duplicate var1) (Duplicate var2) =
+          Just var1 == cast var2
+      go _  _ = False
+
+      _coveredAllCases :: LockstepAction ModelState a -> ()
+      _coveredAllCases = \case
+          New{} -> ()
+          Close{} -> ()
+          Lookups{} -> ()
+          RangeLookup{} -> ()
+          Updates{} -> ()
+          Inserts{} -> ()
+          Deletes{} -> ()
+          RetrieveBlobs{} -> ()
+          Snapshot{} -> ()
+          Open{} -> ()
+          Duplicate{} -> ()
+
+{-------------------------------------------------------------------------------
+  InLockstep
+-------------------------------------------------------------------------------}
+
+instance InLockstep ModelState where
+  type instance ModelOp ModelState = Op
+
+  data instance ModelValue ModelState a where
+    MTableHandle :: Model.TableHandle k v blob -> Val (SUT.TableHandle IO k v blob)
+    MBlobRef :: Model.BlobRef blob -> Val (SUT.BlobRef blob)
+
+    MLookupResult :: Model.C k v blob
+                  => Model.LookupResult k v (Val (SUT.BlobRef blob))
+                  -> Val (SUT.LookupResult k v (SUT.BlobRef blob))
+    MRangeLookupResult :: Model.C k v blob
+                       => Model.RangeLookupResult k v (Val (SUT.BlobRef blob))
+                       -> Val (SUT.RangeLookupResult k v (SUT.BlobRef blob))
+
+    MBlob :: (Show blob, Typeable blob, Eq blob) => Wrap blob -> Val (Wrap blob)
+    MErr :: Model.Err -> Val Model.Err
+
+    MUnit   :: () -> Val ()
+    MPair   :: (Val a, Val b) -> Val (a, b)
+    MEither :: Either (Val a) (Val b) -> Val (Either a b)
+    MList   :: [Val a] -> Val [a]
+
+  data instance Observable ModelState a where
+    OTableHandle :: Obs (SUT.TableHandle IO k v blob)
+    OBlobRef :: Obs (SUT.BlobRef blob)
+
+    OLookupResult :: Model.C k v blob
+                  => Model.LookupResult k v (Obs (SUT.BlobRef blob))
+                  -> Obs (SUT.LookupResult k v (SUT.BlobRef blob))
+    ORangeLookupResult :: Model.C k v blob
+                       => Model.RangeLookupResult k v (Obs (SUT.BlobRef blob))
+                       -> Obs (SUT.RangeLookupResult k v (SUT.BlobRef blob))
```

Comment:
I don't think `OId` would work in this case, because we have to map `SUT.BlobRef` to `Obs (SUT.BlobRef blob)` (see 'OBlobRef').

In this case, I'm assuming that real blob references are handle-like. I'm not sure if it were possible to map real blob references to model blob references if blob references were value-like.